pkg/proxy/iptables/proxier.go

```diff
@@ -1071,6 +1071,7 @@ func (proxier *Proxier) syncProxyRules() {
 				// If/when we support "Local" policy for VIPs, we should update this.
 				writeLine(proxier.natRules, proxier.localDetector.JumpIfNotLocal(args, string(KubeMarkMasqChain))...)
 			}
+			args[1] = string(kubeServicesChain)
```

Comment:
Surya,
This is not an optimal change: The rule is moved into svcChain, but it is not simplified. If a packet hits svcChain, it was matched by the rule in kubeServicesChain already. So there's no point in checking protocol, IP address and port again.

Reply:
> Surya,
> This is not an optimal change: The rule is moved into svcChain, but it is not simplified. If a packet hits svcChain, it was matched by the rule in kubeServicesChain already. So there's no point in checking protocol, IP address and port again.

Hi n0-1, yes I know that it is not optimal and that I am re-checking against protocol and destination. While this may not be necessary for ClusterIP type services alone, however when we have a service of type externalIP or NodePort for example, we still redirect those packets to the same svcChain. So my point was there could be a packet that wasn't matched against that rule and lands in the svcChain. While in theExample:

```
-A KUBE-SERVICES -d 10.96.12.46/32 -p tcp -m comment --comment "default/my-nodeport-service cluster IP" -m tcp --dport 80 -j KUBE-SVC-BD5WSPTUEZLLHMHD
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
-A KUBE-NODEPORTS -p tcp -m comment --comment "default/my-nodeport-service" -m tcp --dport 30007 -j KUBE-SVC-BD5WSPTUEZLLHMHD
-A KUBE-SVC-BD5WSPTUEZLLHMHD ! -s 10.244.0.0/16 -d 10.96.12.46/32 -p tcp -m comment --comment "default/my-nodeport-service cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SVC-BD5WSPTUEZLLHMHD -p tcp -m comment --comment "default/my-nodeport-service" -m tcp --dport 30007 -j KUBE-MARK-MASQ
-A KUBE-SVC-BD5WSPTUEZLLHMHD -m comment --comment "default/my-nodeport-service" -j KUBE-SEP-FPPDB4C4ZRHNV3MY
-A KUBE-SEP-FPPDB4C4ZRHNV3MY -s 10.244.2.4/32 -m comment --comment "default/my-nodeport-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-FPPDB4C4ZRHNV3MY -p tcp -m comment --comment "default/my-nodeport-service" -m tcp -j DNAT --to-destination 10.244.2.4:80
```
I just wanted to be sure we were still doing what we did before even if we moved the rules into the svcChain. I know that in the end we are just masquerading all the packets in the svcChain and then jumping to the sepChain. Probably the comment on the rules is the only thing that is different based on the matches.

Reply:
Yes, the example you pasted seems problematic. I wonder why both packets for 10.96.12.46:80 and <any local address>:30007 are redirected to the svcChain. Is port 30007 somehow linked to that service?

Reply:
> Yes, the example you pasted seems problematic. I wonder why both packets for 10.96.12.46:80 and :30007 are redirected to the svcChain. Is port 30007 somehow linked to that service?

maybe this link: https://groups.google.com/g/kubernetes-sig-network/c/8opCQZhxpxM/m/gCx_DbO-AQAJ can help you understand better, the nodeport service simply builds on top on the clusterip service, that's why both get redirected to the same svcChain, they are not exactly _different_ services in this context.

` ServiceTypeNodePort means a service will be exposed on one port of every node, in addition to 'ClusterIP' type.`
 That's why that svcChain has rules for both the clusterIP and the nodeport. But regardless of that, if we are going to ultimately just masquerade the packets arriving at the svcChain, we don't have to check those fields again since one way or the other they have to get masqueraded and they get filtered already at the kubeServicesChain. I'll anyways add an example test to make the change clear and then ask the reviewers to take a better look.